### ISSUE TITLE
Customized dataset support

### ISSUE THREAD

**Issue by fmggggg** [YKQ98/Matformer#15]
Hi, thank you for providing us this wonderful model. When I tried to train this on my datasets, I encountered some problems. I noticed you only provide the code for benchmarks on Jarvis and MP, could you implement the code for customized datasets where we have cifs and labels? Thanks a lot.

**Comment by fmggggg** [comment]
I've tried to add some code like this, but it can not work.
'''
def convert_structures_to_jarvis_atoms(structure: Structure) -> Atoms:
    return Atoms(
        lattice_mat=structure.lattice.matrix,
        coords=structure.frac_coords,
        elements=[i.symbol for i in structure.species],
        cartesian=False,
    )

class CustomCIFDataset(PygStructureDataset):
    def __init__(self, cif_dir, csv_file, transform=None):
        self.cif_dir = cif_dir
        self.labels = pd.read_csv(csv_file)
        self.transform = transform

    def __len__(self):
        return len(self.labels)

    def __getitem__(self, idx):
        if isinstance(idx, list) or isinstance(idx, np.ndarray):
            idx = idx[0]
        # print(idx)
        cif_filename = self.labels.iloc[idx, 0]
        label = self.labels.iloc[idx, 1]
        cif_path = os.path.join(self.cif_dir, "cifs",f"{cif_filename}.cif")
        structure = Structure.from_file(cif_path, occupancy_tolerance=100.0)
        atoms = convert_structures_to_jarvis_atoms(structure)
        graph = PygGraph.atom_dgl_multigraph(
            atoms, 
            neighbor_strategy="k-nearest", 
            cutoff=8.0, 
            max_neighbors=12
        )
        if self.transform:
            graph = self.transform(graph)
        return graph, label
    
    @staticmethod
    def prepare_batch(batch, device=None, non_blocking=False):
        """Send batched dgl crystal graph to device."""
        g, t = batch
        batch = (
            g.to(device),
            t.to(device, non_blocking=non_blocking),
        )
        return batch
    
def load_splits(cif_dir, labels):
    json_file = os.path.join(cif_dir, 'dataset_split.json')
    with open(json_file, 'r') as f:
        splits = json.load(f)
    
    train_indices = [labels[labels[0] == fn].index[0] for fn in splits['train']]
    val_indices = [labels[labels[0] == fn].index[0] for fn in splits['val']]
    test_indices = [labels[labels[0] == fn].index[0] for fn in splits['test']]
    
    return train_indices, val_indices, test_indices


def get_custom_loaders(cif_dir, csv_file, batch_size, workers=0, pin_memory=True):
    labels = pd.read_csv(csv_file, header=None)  # No header in CSV
    train_indices, val_indices, test_indices = load_splits(cif_dir, labels)
    
    dataset = CustomCIFDataset(cif_dir, csv_file)
    
    train_dataset = Subset(dataset, train_indices)
    val_dataset = Subset(dataset, val_indices)
    test_dataset = Subset(dataset, test_indices)
    
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    
    return train_loader, val_loader, test_loader, dataset.prepare_batch
    
'''

**Comment by fmggggg** [comment]
I've write some code to do this as below, but it can not work.
def convert_structures_to_jarvis_atoms(structure: Structure) -> Atoms:
    return Atoms(
        lattice_mat=structure.lattice.matrix,
        coords=structure.frac_coords,
        elements=[i.symbol for i in structure.species],
        cartesian=False,
    )

class CustomCIFDataset(PygStructureDataset):
    def __init__(self, cif_dir, csv_file, transform=None):
        self.cif_dir = cif_dir
        self.labels = pd.read_csv(csv_file)
        self.transform = transform

    def __len__(self):
        return len(self.labels)

    def __getitem__(self, idx):
        if isinstance(idx, list) or isinstance(idx, np.ndarray):
            idx = idx[0]
        # print(idx)
        cif_filename = self.labels.iloc[idx, 0]
        label = self.labels.iloc[idx, 1]
        cif_path = os.path.join(self.cif_dir, "cifs",f"{cif_filename}.cif")
        structure = Structure.from_file(cif_path, occupancy_tolerance=100.0)
        atoms = convert_structures_to_jarvis_atoms(structure)
        graph = PygGraph.atom_dgl_multigraph(
            atoms, 
            neighbor_strategy="k-nearest", 
            cutoff=8.0, 
            max_neighbors=12
        )
        if self.transform:
            graph = self.transform(graph)
        return graph, label
    
    @staticmethod
    def prepare_batch(batch, device=None, non_blocking=False):
        """Send batched dgl crystal graph to device."""
        g, t = batch
        batch = (
            g.to(device),
            t.to(device, non_blocking=non_blocking),
        )
        return batch
    
def load_splits(cif_dir, labels):
    json_file = os.path.join(cif_dir, 'dataset_split.json')
    with open(json_file, 'r') as f:
        splits = json.load(f)
    
    train_indices = [labels[labels[0] == fn].index[0] for fn in splits['train']]
    val_indices = [labels[labels[0] == fn].index[0] for fn in splits['val']]
    test_indices = [labels[labels[0] == fn].index[0] for fn in splits['test']]
    
    return train_indices, val_indices, test_indices


def get_custom_loaders(cif_dir, csv_file, batch_size, workers=0, pin_memory=True):
    labels = pd.read_csv(csv_file, header=None)  # No header in CSV
    train_indices, val_indices, test_indices = load_splits(cif_dir, labels)
    
    dataset = CustomCIFDataset(cif_dir, csv_file)
    
    train_dataset = Subset(dataset, train_indices)
    val_dataset = Subset(dataset, val_indices)
    test_dataset = Subset(dataset, test_indices)
    
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False, num_workers=workers, pin_memory=pin_memory, collate_fn=CustomCIFDataset.collate)
    
    return train_loader, val_loader, test_loader, dataset.prepare_batch

**Comment by YKQ98** [comment]
I am quite occupied recently, and I suggest you convert these cif files into pymatgen Structure or JARVIS Atoms, and store your interested properties in pandas frame. In this way, you can do your customized dataset by simply changing several lines in data.py.

**Comment by fmggggg** [comment]
Thank you for your reply. That's what I have done, the cifs and properties reading part has been implemented successfully. What confused me was where should I change in data.py to adapt them to imput into the model to train.
